File: flexinput/src/main/java/com/lytefast/flexinput/adapters/PhotoCursorAdapter.java
package com.lytefast.flexinput.adapters;

import android.animation.AnimatorInflater;
import android.animation.AnimatorSet;
import android.content.AsyncQueryHandler;
import android.content.ContentResolver;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.facebook.drawee.view.SimpleDraweeView;
import com.lytefast.flexinput.R;
import com.lytefast.flexinput.model.Photo;
import com.lytefast.flexinput.utils.SelectionCoordinator;


/**
 * {@link android.support.v7.widget.RecyclerView.Adapter} that knows how to load photos from the media store.
 *
 * @author Sam Shih
 */
public class PhotoCursorAdapter extends RecyclerView.Adapter<PhotoCursorAdapter.ViewHolder> {

  private final ContentResolver contentResolver;
  private final SelectionCoordinator<Photo> selectionCoordinator;
  private Cursor cursor;

  private int colId;
  private int colData;
  private int colName;


  public PhotoCursorAdapter(ContentResolver contentResolver,
                            final SelectionCoordinator<Photo> selectionCoordinator) {
    this.contentResolver = contentResolver;
    this.selectionCoordinator = selectionCoordinator.bind(this);

    setHasStableIds(true);
  }

  @Override
  public void onAttachedToRecyclerView(final RecyclerView recyclerView) {
    super.onAttachedToRecyclerView(recyclerView);
    loadPhotos();
  }

  @Override
  public PhotoCursorAdapter.ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
    View view = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.view_grid_image, parent, false);
    return new PhotoCursorAdapter.ViewHolder(view);
  }

  @Override
  public void onBindViewHolder(final PhotoCursorAdapter.ViewHolder holder, final int position) {
    Photo photo = getPhoto(position);
    holder.bind(photo);
  }

  @Override
  public int getItemCount() {
    if (cursor == null) {
      return 0;
    }
    return cursor.getCount();
  }

  @Override
  public long getItemId(final int position) {
    return getPhoto(position).getId();
  }

  @Override
  public void onDetachedFromRecyclerView(final RecyclerView recyclerView) {
    cursor.close();
    super.onDetachedFromRecyclerView(recyclerView);
  }

  @Nullable
  public void loadPhotos() {
    final AsyncQueryHandler asyncQueryHandler = new AsyncQueryHandler(contentResolver) {
      @Override
      protected void onQueryComplete(final int token, final Object cookie, final Cursor cursor) {
        PhotoCursorAdapter.this.colId = cursor.getColumnIndex(MediaStore.Images.Media._ID);
        PhotoCursorAdapter.this.colData = cursor.getColumnIndex(MediaStore.Images.Media.DATA);
        PhotoCursorAdapter.this.colName = cursor.getColumnIndex(MediaStore.Images.Media.DISPLAY_NAME);

        PhotoCursorAdapter.this.cursor = cursor;
        notifyDataSetChanged();
      }
    };
    asyncQueryHandler.startQuery(1, this,
        MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
        new String[]{
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.DISPLAY_NAME},
        null, null, MediaStore.Images.Media.DATE_ADDED + " DESC");
  }

  private Photo getPhoto(int position) {
    cursor.moveToPosition(position);
    final String photoId = Integer.toString(cursor.getInt(colId));
    final Uri fileUri = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, photoId);
    return new Photo(
        cursor.getLong(colId), fileUri, cursor.getString(colName), cursor.getString(colData));
  }

  protected class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
    private final AnimatorSet shrinkAnim;
    private final AnimatorSet growAnim;

    public final SimpleDraweeView imageView;
    private final View checkIndicator;

    private Photo photo = null;


    public ViewHolder(final View itemView) {
      super(itemView);
      this.itemView.setOnClickListener(this);

      this.imageView = (SimpleDraweeView) itemView.findViewById(R.id.content_iv);
      this.checkIndicator = itemView.findViewById(R.id.item_check_indicator);

      //region Perf: Load animations once
      this.shrinkAnim = (AnimatorSet) AnimatorInflater.loadAnimator(
          itemView.getContext(), R.animator.selection_shrink);
      this.shrinkAnim.setTarget(imageView);

      this.growAnim = (AnimatorSet) AnimatorInflater.loadAnimator(
          itemView.getContext(), R.animator.selection_grow);
      this.growAnim.setTarget(imageView);
      //endregion
    }

    public void bind(final Photo photo) {
      this.photo = photo;
      setSelected(selectionCoordinator.isSelected(photo, getAdapterPosition()), false);

      Uri thumbnailUri = photo.getThumbnailUri(contentResolver);
      if (thumbnailUri == null) {
        imageView.setImageURI((String) null);
      } else {
        imageView.setImageURI(thumbnailUri);
      }
    }

    void setSelected(boolean isSelected, boolean isAnimationRequested) {
      itemView.setSelected(isSelected);

      if (isSelected) {
        checkIndicator.setVisibility(View.VISIBLE);
        if (imageView.getScaleX() == 1.0f) {
          shrinkAnim.start();
          if (!isAnimationRequested) {
            shrinkAnim.end();
          }
        }
      } else {
        checkIndicator.setVisibility(View.GONE);
        if (imageView.getScaleX() != 1.0f) {
          growAnim.start();
          if (!isAnimationRequested) {
            growAnim.end();
          }
        }
      }
    }

    @Override
    public void onClick(final View v) {
      setSelected(selectionCoordinator.toggleItem(photo, getAdapterPosition()), true);
    }
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/adapters/PhotoCursorAdapter.kt
package com.lytefast.flexinput.adapters

import android.animation.AnimatorInflater
import android.animation.AnimatorSet
import android.content.AsyncQueryHandler
import android.content.ContentResolver
import android.content.Context
import android.database.Cursor
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.CancellationSignal
import android.provider.MediaStore
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.facebook.drawee.drawable.FadeDrawable
import com.facebook.drawee.drawable.ScalingUtils
import com.facebook.drawee.generic.RoundingParams
import com.facebook.drawee.view.SimpleDraweeView
import com.lytefast.flexinput.R
import com.lytefast.flexinput.model.Photo
import com.lytefast.flexinput.utils.BuildUtils
import com.lytefast.flexinput.utils.SelectionCoordinator
import com.lytefast.flexinput.utils.ThumbnailUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch


/**
 * [RecyclerView.Adapter] that knows how to load photos from the media store.
 *
 * @author Sam Shih
 */
class PhotoCursorAdapter(private val contentResolver: ContentResolver,
                         selectionCoordinator: SelectionCoordinator<*, Photo>,
                         val thumbnailWidth: Int,
                         val thumbnailHeight: Int)
  : RecyclerView.Adapter<PhotoCursorAdapter.ViewHolder>() {
  private val selectionCoordinator: SelectionCoordinator<*, Photo> = selectionCoordinator.bind(this)
  private var cursor: Cursor? = null

  private var colId: Int = 0
  private var colData: Int = 0
  private var colName: Int = 0


  init {
    setHasStableIds(true)
  }

  override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
    super.onAttachedToRecyclerView(recyclerView)

    emptyColorDrawable = ColorDrawable(recyclerView.context.themeColor(R.attr.flexInputDialogBackground))
    loadPhotos()
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    val view = LayoutInflater.from(parent.context)
        .inflate(R.layout.view_grid_image, parent, false)
    return ViewHolder(view)
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    val photo = this[position]
    holder.bind(photo)
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: MutableList<Any>) {
    payloads
        .firstOrNull { it is SelectionCoordinator.SelectionEvent<*> }
        ?.let { it as? SelectionCoordinator.SelectionEvent<*> }
        ?.also {
          holder.setSelected(it.isSelected, isAnimationRequested = true)
          return
        }
    super.onBindViewHolder(holder, position, payloads)
  }

  override fun onViewRecycled(holder: ViewHolder) {
    super.onViewRecycled(holder)
    holder.onViewRecycled()
  }

  override fun getItemCount(): Int = cursor?.count ?: 0

  override fun getItemId(position: Int): Long = this[position]?.id ?: -1

  override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
    cursor?.close()
    super.onDetachedFromRecyclerView(recyclerView)
  }

  fun loadPhotos() {
    class LoadQueryHandler : AsyncQueryHandler(contentResolver) {
      override fun onQueryComplete(token: Int, cookie: Any, cursor: Cursor?) {
        if (cursor == null) {
          return
        }
        this@PhotoCursorAdapter.apply {
          this.colId = cursor.getColumnIndex(MediaStore.Images.Media._ID)
          this.colData = cursor.getColumnIndex(MediaStore.Images.Media.DATA)
          this.colName = cursor.getColumnIndex(MediaStore.Images.Media.DISPLAY_NAME)
          this.cursor = cursor
        }
        notifyDataSetChanged()
      }
    }

    LoadQueryHandler().startQuery(1, this,
        MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
        arrayOf(MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.DISPLAY_NAME),
        null, null, "${MediaStore.Images.Media.DATE_ADDED} DESC")
  }

  private operator fun get(position: Int): Photo? =
      cursor?.let {
        it.moveToPosition(position)
        val photoId = it.getLong(colId)
        val fileUri = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, photoId.toString())
        Photo(
            id = photoId,
            uri = fileUri,
            displayName = it.getString(colName) ?: "img-$photoId",
            photoDataLocation = it.getString(colData))
      }

  inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
    private val imageView: SimpleDraweeView = itemView.findViewById(R.id.content_iv)
    private val checkIndicator: SimpleDraweeView = itemView.findViewById(R.id.item_check_indicator)

    private var photo: Photo? = null
    private var loadThumbnailJob: Job? = null
    private var thumbnailBitmap: Bitmap? = null
    private var thumbnailDrawable: BitmapDrawable? = null
    private var holderFadeDrawable: FadeDrawable? = null
    private var cancelGetThumbnailSignal: CancellationSignal? = null

    private val shrinkAnim: AnimatorSet
    private val growAnim: AnimatorSet

    init {
      this.itemView.setOnClickListener(this)

      //region Perf: Load animations once
      this.shrinkAnim = AnimatorInflater.loadAnimator(
          imageView.context, R.animator.selection_shrink) as AnimatorSet
      this.shrinkAnim.setTarget(imageView)

      this.growAnim = AnimatorInflater.loadAnimator(
          itemView.context, R.animator.selection_grow) as AnimatorSet
      this.growAnim.setTarget(imageView)
      //endregion
    }

    fun bind(photo: Photo?) {
      this.photo = photo

      if (photo != null) {
        setSelected(selectionCoordinator.isSelected(photo, adapterPosition), false)
      }

      if (BuildUtils.isAndroidQ()) {
        clear()

        imageView.hierarchy.setPlaceholderImage(emptyColorDrawable, ScalingUtils.ScaleType.CENTER)
        cancelGetThumbnailSignal = CancellationSignal()

        // Ensure this executes on the main thread for UI interaction (as opposed to IO)
        loadThumbnailJob = GlobalScope.launch(context = Dispatchers.Main) {
          val thumbnailUri = photo?.uri ?: return@launch
          thumbnailBitmap = ThumbnailUtils.getThumbnailQ(
              contentResolver, thumbnailUri,
              thumbnailWidth, thumbnailHeight,
              cancelGetThumbnailSignal)

          thumbnailDrawable = BitmapDrawable(imageView.resources, thumbnailBitmap)
          val fadeDrawable = FadeDrawable(arrayOf(emptyColorDrawable, thumbnailDrawable))

          fadeDrawable.transitionDuration = 300
          val roundingParams = RoundingParams.fromCornersRadius(imageView.context.dpToPixels(4f))
          roundingParams.overlayColor = imageView.context.themeColor(R.attr.flexInputDialogBackground)
          imageView.hierarchy.roundingParams = roundingParams
          imageView.hierarchy.setPlaceholderImage(fadeDrawable, ScalingUtils.ScaleType.CENTER_CROP)
          fadeDrawable.fadeToLayer(1)
          holderFadeDrawable = fadeDrawable
        }
      } else {
        val thumbnailUri = photo?.getThumbnailUri(contentResolver)
        imageView.setImageURI(thumbnailUri, imageView.context)
      }
    }

    fun setSelected(isSelected: Boolean, isAnimationRequested: Boolean = true) {
      itemView.isSelected = isSelected

      fun scaleImage(animation: AnimatorSet) {
        animation.start()
        if (!isAnimationRequested) {
          animation.end()
        }
      }

      if (isSelected) {
        checkIndicator.visibility = View.VISIBLE
        if (imageView.scaleX == 1.0f) scaleImage(shrinkAnim)
      } else {
        checkIndicator.visibility = View.GONE
        if (imageView.scaleX != 1.0f) scaleImage(growAnim)
      }
    }

    override fun onClick(v: View) {
      selectionCoordinator.toggleItem(photo, adapterPosition)
    }

    fun onViewRecycled() {
      clear()
    }

    private fun clear() {
      if (BuildUtils.isAndroidQ()) {
        loadThumbnailJob?.cancel()
        cancelGetThumbnailSignal?.cancel()

        thumbnailDrawable?.bitmap?.recycle()
        thumbnailDrawable = null
        thumbnailBitmap?.recycle()
        thumbnailBitmap = null
        val fadeBitmapDrawable = holderFadeDrawable?.getDrawable(1) as? BitmapDrawable
        fadeBitmapDrawable?.bitmap?.recycle()
        holderFadeDrawable = null
      }
    }
  }

  companion object {
    private var emptyColorDrawable: Drawable? = null

    fun Context.dpToPixels(dipValue: Float) =
        TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dipValue, resources.displayMetrics)

    fun Context.themeColor(themeAttributeId: Int): Int {
      val outValue = TypedValue()
      val wasResolved = theme.resolveAttribute(themeAttributeId, outValue, true)
      return if (wasResolved) {
        ContextCompat.getColor(this, outValue.resourceId)
      } else {
        0
      }
    }
  }
}
